### ISSUE TITLE
Some questions about the baseline accuracy in the paper

### ISSUE THREAD

**Issue by SouGongngai** [microsoft/SGN#4]
### Firstly, thank very much for your contribution. Your paper gave me a lot of inspiration.
But I still have some questions. I ran the source code again in my environment, but I can't achieve the same accuracy especially in NTU_RGBD dataset. My environment is one GeForce RTX 2080 Ti, Python is 3.6.5, Pytorch is 1.2.0 and  I use python virtual environment instead of Anaconda.
Here is my experiment results:
![屏幕快照 2020-06-22 下午7 51 49](https://user-images.githubusercontent.com/11470165/85284456-d1d05280-b4c1-11ea-9885-2a47731a97c2.png)
## Please🙏, your replay is very important to me.

**Comment by 541968679** [comment]
> ### Firstly, thank very much for your contribution. Your paper gave me a lot of inspiration.
> But I still have some questions. I ran the source code again in my environment, but I can't achieve the same accuracy especially in NTU_RGBD dataset. My environment is one GeForce RTX 2080 Ti, Python is 3.6.5, Pytorch is 1.2.0 and I use python virtual environment instead of Anaconda.
> Here is my experiment results:
> ![屏幕快照 2020-06-22 下午7 51 49](https://user-images.githubusercontent.com/11470165/85284456-d1d05280-b4c1-11ea-9885-2a47731a97c2.png)
> 
> ## Please🙏, your replay is very important to me.

hello, i have the same question as you, i checked all factor which may have effect on the final performance but still cannot get the results described in the article. but i also tested another earlier work,. i think we can communicate with each other. you can add my QQ(541968679) or my wechat(h541968679) if you don't mind. thank you very much!

**Comment by SouGongngai** [comment]
👌

**Comment by shuidongliu** [comment]
> ### Firstly, thank very much for your contribution. Your paper gave me a lot of inspiration.
> But I still have some questions. I ran the source code again in my environment, but I can't achieve the same accuracy especially in NTU_RGBD dataset. My environment is one GeForce RTX 2080 Ti, Python is 3.6.5, Pytorch is 1.2.0 and I use python virtual environment instead of Anaconda.
> Here is my experiment results:
> ![屏幕快照 2020-06-22 下午7 51 49](https://user-images.githubusercontent.com/11470165/85284456-d1d05280-b4c1-11ea-9885-2a47731a97c2.png)
> 
> ## Please🙏, your replay is very important to me.

The version of cuda (9.2 or higher) is important.

**Comment by 541968679** [comment]
> > ### Firstly, thank very much for your contribution. Your paper gave me a lot of inspiration.
> > But I still have some questions. I ran the source code again in my environment, but I can't achieve the same accuracy especially in NTU_RGBD dataset. My environment is one GeForce RTX 2080 Ti, Python is 3.6.5, Pytorch is 1.2.0 and I use python virtual environment instead of Anaconda.
> > Here is my experiment results:
> > ![屏幕快照 2020-06-22 下午7 51 49](https://user-images.githubusercontent.com/11470165/85284456-d1d05280-b4c1-11ea-9885-2a47731a97c2.png)
> > ## Please🙏, your replay is very important to me.
> 
> The version of cuda (9.2 or higher) is important.

hello, i checked my cuda version and it is 10.2.89，but i still cannot get right result. Could you please share your environment config if you can reproduce the result in the paper?

**Comment by SouGongngai** [comment]
The version of my cuda is 10.1, should I upgrade to 10.2?

**Comment by shuidongliu** [comment]
> > > ### Firstly, thank very much for your contribution. Your paper gave me a lot of inspiration.
> > > But I still have some questions. I ran the source code again in my environment, but I can't achieve the same accuracy especially in NTU_RGBD dataset. My environment is one GeForce RTX 2080 Ti, Python is 3.6.5, Pytorch is 1.2.0 and I use python virtual environment instead of Anaconda.
> > > Here is my experiment results:
> > > ![屏幕快照 2020-06-22 下午7 51 49](https://user-images.githubusercontent.com/11470165/85284456-d1d05280-b4c1-11ea-9885-2a47731a97c2.png)
> > > ## Please🙏, your replay is very important to me.
> > 
> > 
> > The version of cuda (9.2 or higher) is important.
> 
> hello, i checked my cuda version and it is 10.2.89，but i still cannot get right result. Could you please share your environment config if you can reproduce the result in the paper?

I graduated one month ago.  Therefore, I can only provide you the environment config based on my memory. The version of Python is 3.6 (Anaconda), the version of Pytorch is 1.3, the version of cuda is 10.0.

**Comment by SouGongngai** [comment]
Thank you very much for your replay! Recently I upgraded the Pytorch from 10.2 to 10.3. And Xsub can achieve the accuracy of 87.50% but the accuracy of Xview is still 94.221%😂. Next I will re-test it according to your environment and I hope to succeeded this time🙏.

**Comment by SouGongngai** [comment]
Thank for your help! I upgrade my Pytorch version from 1.2 to 1.3 and it is really useful😄.

**Comment by desertfex** [comment]
@SouGongngai  Do you mean that your Xsub can achieve the accuracy of 89.00% after upgrade the pytorch version form 1.2 to 1.3?

**Comment by SouGongngai** [comment]
Yes, but you need to try a few more times.

**Comment by desertfex** [comment]
@SouGongngai Thank you. But what is the meaning of "a few more times"? Do you mean more epoches? If I want to train more epoches, what should I do?  Will change the "max_epochs" in main.py works? It seems that I change the "max_epochs" in main.py  from 120 to 200, the training process still ended by 120.  What should I do? Thank you.

**Comment by desertfex** [comment]
@SouGongngai  My fault. I can train more epoches now. But I cannot achieve the accuracy. How many epoches you train to achieve 89.00%?

**Comment by SouGongngai** [comment]
You are welcome. My epoch is still 120. I mean it may be necessary to do more experiments, because there may be a fluctuation between the results of each time, which is about 0.3.

**Comment by desertfex** [comment]
Thank you for your reply. But the random seed is fixed in main.py:
np.random.seed(1337)
So why the result is different for every try? Do you have any idea?

**Comment by binli-92** [comment]
> You are welcome. My epoch is still 120. I mean it may be necessary to do more experiments, because there may be a fluctuation between the results of each time, which is about 0.3.

So you mean we can reach a low limit of 88.5% for NTU60 CS every trail? I can only get 87%+.

**Comment by zheda-libin** [comment]
> You are welcome. My epoch is still 120. I mean it may be necessary to do more experiments, because there may be a fluctuation between the results of each time, which is about 0.3.

@SouGongngai  So you mean we can at least reach 88%+ for NTU60 CS for each trail? I can only get 87%+.

**Comment by SouGongngai** [comment]
According to the environment given by the author, I retest the code.Slightly different, but small.

> > > > ### Firstly, thank very much for your contribution. Your paper gave me a lot of inspiration.
> > > > But I still have some questions. I ran the source code again in my environment, but I can't achieve the same accuracy especially in NTU_RGBD dataset. My environment is one GeForce RTX 2080 Ti, Python is 3.6.5, Pytorch is 1.2.0 and I use python virtual environment instead of Anaconda.
> > > > Here is my experiment results:
> > > > ![屏幕快照 2020-06-22 下午7 51 49](https://user-images.githubusercontent.com/11470165/85284456-d1d05280-b4c1-11ea-9885-2a47731a97c2.png)
> > > > ## Please🙏, your replay is very important to me.
> > > 
> > > 
> > > The version of cuda (9.2 or higher) is important.
> > 
> > 
> > hello, i checked my cuda version and it is 10.2.89，but i still cannot get right result. Could you please share your environment config if you can reproduce the result in the paper?
> 
> I graduated one month ago. Therefore, I can only provide you the environment config based on my memory. The version of Python is 3.6 (Anaconda), the version of Pytorch is 1.3, the version of cuda is 10.0.

According to the environment given by the author, I retest the code. Here are the results. Almost the same, except for the results of Xsub(NTU_RGBD_A60).
![屏幕快照 2020-06-28 上午10 41 14](https://user-images.githubusercontent.com/11470165/85936402-18440800-b92d-11ea-9899-1d950c49a2f9.png)
![屏幕快照 2020-06-28 上午10 41 48](https://user-images.githubusercontent.com/11470165/85936411-37429a00-b92d-11ea-82d4-ca2fc26a2c92.png)
![屏幕快照 2020-06-28 上午10 44 22](https://user-images.githubusercontent.com/11470165/85936413-390c5d80-b92d-11ea-8aa9-b22e18dd8cab.png)
![屏幕快照 2020-06-28 上午10 45 15](https://user-images.githubusercontent.com/11470165/85936415-39a4f400-b92d-11ea-9e81-ef49ae6a8853.png)
![屏幕快照 2020-06-28 上午10 45 24](https://user-images.githubusercontent.com/11470165/85936416-3a3d8a80-b92d-11ea-8480-0465251497f3.png)

**Comment by desertfex** [comment]
The pictures are not showed.

**Comment by 541968679** [comment]
you can set: 
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False
and can got 88.750 for xsub

**Comment by desertfex** [comment]
@SouGongngai 
 The authors only provide the codes to get the NTU60 data. But how to get the NTU120 data?  I will grateful if you teach me or could you please upload your codes? Thank you very much.

**Comment by 541968679** [comment]
话说做这个领域的全都是中国人，你们还用什么英文嘛
我在下面的配置是可以还原出ntu60 xsub 89.0的结果：
python: 3.6.10(anaconda)
显卡：单个RTX2080
驱动：440.33.01
cuda: 10.2.89
cudnn: 765

此外还需要设置：
torch.backends.cudnn.deterministic = True
 torch.backends.cudnn.benchmark = False

**Comment by 541968679** [comment]
作者没有提供NTU120数据集的相关处理代码，但自己写起来也并不难，只需要从官网下载NTU120数据集然后与NTU60合并，接着重新生成statistics目录下的txt文件，最后运行数据处理代码即可，嫌麻烦可以找我要，微信在上面。

个人认为这篇文章还是非常具有研讨和跟进意义的，主要有以下几个点：
1. 网络结构轻量化，这是最明显的一个点。但其实作者没有明说能让网络轻量化的关键点在哪，
我个人通过实验发现数据的预处理是最关键的点，作者将原始的不规则长度的样本通过下采样的统一采样为长度为20的样本，这在很大程度上降低了网络对时域感受野的要求，而时域感受野与参数量有直接关系，一个最明显的对比例子就是基于ST-GCN的那些工作，他们都采用了9层网络，并且统一将序列填补到300帧，但这篇文章证明GCN完全不用堆叠9层，只需要三层（甚至两层）也能达到相同的效果，不管是训练速度还是网络规模都远优于基于STGCN的工作（例如2S-AGCN，以及今年的其他几篇CVPR）。我个人认为这是这篇文章最重要的贡献。
2.第二个最重要的贡献是网络直接使用注意力机制对关节关系进行建模，实际上从这个层面来讲就与GCN无关了。
3.第三个才是网络提出的semantic编码，其实就是位置编码，并且用不用one-hot都无所谓。

**Comment by 541968679** [comment]
对了他这种时序抽样的训练和测试模式其实也是有缺点的，在NTU120上的表现就明显不如2S-AGCN，当然也可能是网络体量的原因

**Comment by pudding126** [comment]
> ### Firstly, thank very much for your contribution. Your paper gave me a lot of inspiration.
> But I still have some questions. I ran the source code again in my environment, but I can't achieve the same accuracy especially in NTU_RGBD dataset. My environment is one GeForce RTX 2080 Ti, Python is 3.6.5, Pytorch is 1.2.0 and I use python virtual environment instead of Anaconda. Here is my experiment results: ![屏幕快照 2020-06-22 下午7 51 49](https://user-images.githubusercontent.com/11470165/85284456-d1d05280-b4c1-11ea-9885-2a47731a97c2.png)
> 
> ## Please🙏, your replay is very important to me.

您好，请问您的代码中有’./data/ntu/NTU_CS.h5‘这个文件吗
